app/lib/proxy_api/salt.rb

```diff
@@ -5,24 +5,18 @@ def initialize(args)
       super args
     end
 
-    def autosign_list
-      parse(get('autosign'))
+    def autosign_create(serial)
+      parse(post('', "autosign/#{URI.escape(serial)}"))
     rescue => e
-      raise ProxyException.new(url, e, N_('Unable to fetch autosign list'))
+      raise ProxyException.new(url, e, N_('Unable to create Salt autosign with serial %s'), serial)
     end
 
-    def autosign_create(name)
```

Comment:
if I understand the code correctly, does this also remove the api calls to create autosign entries? Even though autosign.conf won't be the default anymore, we are still using this in our environment on specific salt proxies (ie. ones that really don't provide provisioning but just allow IOT devices that are being flashed to appear in foreman).
So I'd appreciate if at least the api calls for creating autosign records would be available after the change :)

Reply:
Thanks for your feedback!
Just to understand the use case: You are talking about the API endpoint on the Smart Proxy side, aren't you? So, you want to create an autosign key on your own and send it to the Smart Proxy instead of going the way through the host provisioning on `foreman_salt` side. I think this is usually the case when you want to add an already existing host for example.
The API endpoint on the Smart Proxy side still exists and is used by `foreman_salt` to transmit a key to the Smart Proxy as well.

Nevertheless, I must confess that I didn't think that use case through: If you create a key on your own and transmit it to the Smart Proxy, we run into an error when the key shall be deleted afterwards. The Smart Proxy will trigger the key deletion process via an HTTPS request to the Foreman as soon as the Minion was authenticated for the first time. But, the Foreman doesn't know the key since it was created by another entity and sent to the Smart Proxy. Therefore, `foreman_salt` will fail when it wants to call the Smart Proxy API to delete the key. In order to do this, `foreman_salt` needs to know the autosign key as well.

What do you think about providing an API entry within `foreman_salt` to enable keys which were generated by another entity and _not_ during the provisioning process?

Reply:
I think I commented the wrong part of the code, sorry :)
https://github.com/theforeman/foreman_salt/pull/136/files#diff-4b3c1e928bb6ed000f02e5c9b08f46a2638fbda55bf1bf5f4a186e84179ef3c3
This is what I meant. So I am talking about the actual foreman api and the possibility to create autosign entries directly via foreman API.
Our provisioning procedure for these devices is:
* Run scripts to create the host in foreman and other systems, including the correct host parameters. These scripts also create an autosign entry for the hostname to be connected to foreman (yes, hostname is insecure anyway, but in this usecase it'll do just fine)
* Flash device with predefined image, so there's no way to create a grains key in that procedure to be injected into the image
* Configure system, the image already has the config to talk to the correct foreman proxy/saltmaster
The API calls in foreman_salt are removed by that PR mentioned above...

Reply:
Alright, makes sense to me. According to your procedure, you still have to add the key kind of manually to the host since you have your predefined image without the grains key, as you mentioned. I hope that will work for you somehow.

I took a look at the foreman API as it existed before. It takes only the Smart Proxy ID and the autosign key (namely `record`) which shall be added. Unfortunately, we need to associate the autosign key with the host it corresponds to. Otherwise, we won't be able to tell the status of a host (is it already accepted or not?).
So, is it okay for you to add another parameter like "host_name" to the API, in order to associate the given key with a specific host?

Reply:
I think you may need a new procedure. The main reason we want to implement a new authentication method are more restrictive permissions on the `autosign.conf` file with a new Salt version. You can read about it [here](https://docs.saltproject.io/en/latest/ref/configuration/master.html#autosign-file). Since no other user besides the `saltuser` has write-access on that file any more, we cannot add new host entries with our `foreman-proxy` user.

Nevertheless, I like your proposal with an API endpoint to enable the creation of autosign keys via API. I would suggest the following two additions:
* Implement `/salt_autosign/create/:host_name` to generate a new autosign key for the given host and transmit it to the Smart Proxy. Send the generated autosign key in the answer of the request such that the initiator of the request knows the key as well.
* Implement `/salt_autosign/:host_name` which needs a parameter `autosign_key` in the request. Instead of generating the key by itself, it takes a previously generated "third-party" key and just transmits this one to the Smart Proxy.

By default, the first API endpoint should be preferred, since it takes advantage of generating the autosign key which makes it less likely to have duplicate keys. Still, in case someone wants to use their own key, they can use the second endpoint.
Moreover, we need `:host_name` instead of `:smart_proxy_id` because the autosign key must be associate with a host in order to update the host status (so we can say "Host has/hasn't been accepted by the Salt master").

Do you think you can manage to implement a new procedure since `autosign.conf` is somehow deprecated for us? And what to you think about the two API endpoints I just suggested?

Reply:
> I think you may need a new procedure. The main reason we want to implement a new authentication method are more restrictive permissions on the `autosign.conf` file with a new Salt version. You can read about it [here](https://docs.saltproject.io/en/latest/ref/configuration/master.html#autosign-file). Since no other user besides the `saltuser` has write-access on that file any more, we cannot add new host entries with our `foreman-proxy` user.

That's not entirely true, I actually have exactly that setup working on salt 3003 :-)
The only thing that needs to be confiured is that the user running salt needs to be in the foreman-proxy group and the configuration option permissive_pki_access: True

Reply:
You are right, I didn't consider that. So, I will make the Autosign via Grains procedure also compatible with `autosign.conf` & test it with my setup.

I'm going to inform you when it's done, thanks for the advice :)